### ISSUE TITLE
Batchnorm vs dropout

### ISSUE THREAD

**Issue by AlekzNet** [jcjohnson/torch-rnn#62]
I wonder if anybody has made any comparisons between batch normalization and dropout, and/or a combination of them. I've seen some papers on that, they state that for text processing the difference is not as pronounced as, for example, for images, but I did not see any proper comparisons.

**Comment by jcjohnson** [comment]
I didn't do any super controlled experiments, but deeper models with batchnorm do seem to train faster.

**Comment by AlekzNet** [comment]
I made some comparisons for 3x128 model (16MB corpus, seq_length=400, batch_size=200, 200 iteration per epoch):

![3x128-comparison](https://cloud.githubusercontent.com/assets/10780668/15079741/a6b65272-13ba-11e6-9bda-2b8834769600.png)

The peaks at iteration 400 are related to the learning rate decay (0.92) for the first 3 test runs, I removed the decay for all subsequent runs.

The bottom line - the batch normalization DOES help. E.g. without the BN, the highest possible LR was 0.01, (0.05 with the BN).   "Light" dropout (Do=0.1) helps if there is no BN. 

Fun! ;)

**Comment by jcjohnson** [comment]
Those are some beautiful curves - thanks for sharing! It's interesting that the models without BN show a plateau at the beginning - that suggests to me that maybe the initialization is not correct.

**Comment by AlekzNet** [comment]
I wonder if adding the gradient noise can help with bad initializations: https://github.com/jcjohnson/torch-rnn/issues/85 It should help in such cases. But the BN does its magic perfectly! 

Another interesting thing is the LR choice. There must be an optimum for each model size (looking for related articles at the moment ;) ). For this particular model (3x128) 0.01 is clearly better than 0.05 and 0.002 (the default for torch-rnn and char-rnn).
